platform/api-server/api/controllers/platform-api-controller.js

```diff
@@ -99,3 +103,244 @@ exports.updatePrivacyPolicy = async function(req, res) {
+
+    // Obtain gateway resource information to make a scheduling decision.
+    const graph = await utils.getLinkGraph();
+    const gatewayResources = await Promise.all(Object.keys(graph.data).map((gatewayId) => {
+        const gatewayInfo = graph.data[gatewayId];
+        const ip = gatewayInfo.ip;
+
+        // Get a count of devices by type for each gateway.
+        // This is used for deployment metadata that specifies a particular kind of device.
+        var deviceTypes = new Map();
+        gatewayInfo.devices.forEach((device) => {
+            if (deviceTypes.has(device.type)) {
+                deviceTypes[device.type] += 1;
+            } else {
+                deviceTypes.set(device.type, 1);
+            }
+        });
+
+        const opts = {
+            method: 'GET',
+            uri: `http://${ip}:5000/gateway/resources`,
+            json: true
+        };
+
+        return request(opts)
+            .then((resources) => {
+                const deviceInfo = {
+                    id: gatewayId,
+                    ip: ip,
+                    resources: resources,
+                    deviceIds: gatewayInfo.devices.map(device => device.id),
+                    deviceTypes: deviceTypes
+                };
+
+                return deviceInfo;
+            });
+    }));
+
+    // Run the scheduling algorithm to determine where to put the application.
+    const gateway = schedule(deployMetadata, runMetadata, gatewayResources);
+    if (gateway !== null) {
+        console.log(`Sending application to run on ${gateway.ip}.`);
+
+        const gatewayUri = `http://${gateway.ip}:5000/gateway/execute-app-v2`;
+        const formData = {
+            'appPackage': fsExtra.createReadStream(packagePath),
+            'deployMetadata': fsExtra.createReadStream(deployMetadataPath)
+        };
+        const opts = {
+            method: 'POST',
+            uri: gatewayUri,
+            formData: formData
+        };
+
+        request(opts)
+            .then(() => { res.sendStatus(204); },
+                  () => { res.sendStatus(500); });
+    } else {
+        // No gateways available for the application.
+        // Send back 503 Service Unavailable.
+        res.sendStatus(503);
+    }
+};
+
+/** Decide which gateway should run an application.
+ *
+ * Makes a decision to schedule an application on a gateway provided in `gateways`.
+ * If no gateway is suitable for the application, this function returns null.
+ *
+ * @param deployMetadata Deploy-time application information.
+ * @param runMetadata Build-time application information.
+ */
+function schedule(deployMetadata, runMetadata, gateways) {
+    // (1) Remove gateways based on specs, requirements, and connected devices.
+    // Remove overloaded gateways.
+    // Inspect CPU and memory load and removes gateways that are above the threshold.
+    var candidates = gateways.filter((gw) => {
+        return gw.resources.cpuFreePercent < 80
```

Comment:
create global variables for the thresholds for better readability + any future updates